src/compiler/checker.ts

```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {

```

Comment:
The AST has a single root type (`TypePredicateNode`) with _two_ possible types for the `parameterName` member - `Identifier` or `ThisTypeNode` (could just as well be `ThisKeyword`, but a `ThisTypeNode` seems more appropriate). The current `TypePredicate` hierarchy reflects that - the base type can be either form, with `IdentifierTypePredicate` being specialized to the `Identifier` case and `ThisTypePredicate` being specialized for the `ThisTypeNode` case, since `TypePredicate` is structured via inheritance rather than composition (and refactoring it to match the grammar and be structured by composition would change a bunch code and be harder to adapt to by API consumers).

And they _are_ that different - in every case where we handle a type predicate in the checker (all two of them) there's a switch delegating to different behaviors for each kind.